Review of Localzone

omus opened the pull request.

Michael-Klassen committed:
Created get_localzone funtion for UNIX and OS X

Michael-Klassen committed:
Create unit tests for local

The testcases for local are written such that they temporarily
overide functions such that we don't have to mess with creating
files for testing. Since Julia is trying to optimize compilation of
methods we can end up in situations where we overwrite a method
used inside of another method and it continues to use the original
method. Using `[]...` allows us to change the behaviour of a
function.

See https://github.com/JuliaLang/julia/issues/265 for details on the
core issue.

Michael-Klassen committed:
Add vectorized versions of accessors

Michael-Klassen committed:
Add TimeZones.now method

Similar to Base.now but returns a ZonedDateTime. Note that we are not
using Base.now since that returns a local system time zone DateTime and
it is preferrable to use a UTC DateTime.

omus committed:
Switched to a stripped down of Fixtures.

The Fixtures package has yet to be updated for Julia 0.4 and even when
it has update it appears doubtful it will work correctly due to:

https://github.com/JuliaLang/julia/issues/265

As a temporarly solution the fundamentals of Fixtures have been
included for running the test suite.

omus commented on .gitignore:
```diff
@@ -5,3 +5,4 @@
 # Ignore dependencies that are build/retrieved
 deps/tzdata
 deps/compiled
+deps/windows
```
Should include a blank newline at the end of files.

omus commented on deps/build.jl:
```diff
@@ -1,4 +1,5 @@
-import TimeZones: TZDATA_DIR, COMPILED_DIR
+using LightXML
```
Include in any used packages in the REQUIRE file. In this case should be `@windows LightXML` in the REQUIRE file and here it should be `@windows_only using LightXML`

omus commented on deps/build.jl:
```diff
@@ -43,3 +44,60 @@ end
+isdir(WINDOWS_DIR) || mkdir(WINDOWS_DIR)
+
+# Windows is weird and uses its own timezone
+info("Downloading Windows to TZ data")
```
Should all be included in a `OS_NAME == :Windows` conditional block

omus commented on deps/build.jl:
```diff
@@ -43,3 +44,60 @@ end
+            else
+                rethrow()
+            end
+        end
```
You shouldn't need all this logic for downloading this file. It is only needed above to deal with the low availability IANA FTP server.

omus commented on deps/build.jl:
```diff
@@ -43,3 +44,60 @@ end
+# tzdata/Olsen timezone names, by retrieving a file.
+if 1==1
+    xml_source = "http://unicode.org/cldr/data/common/supplemental/windowsZones.xml"
+    xml_file = joinpath(WINDOWS_DIR, "windowsZones.xml")
```
Maybe just have this in the `deps` folder? Why does this need a brand new directory?

omus commented on deps/build.jl:
```diff
@@ -43,3 +44,60 @@ end
+    win_tz = Dict{String,String}()
+
+    for mapzone in mapZones
+        if attribute(mapzone, "territory") == "001"
```
What's up with this? Needs a comment. Explain what you are trying to do here.

omus commented on src/timezones/conversions.jl:
```diff
@@ -0,0 +1,12 @@
+
+function now()
+    return ZonedDateTime(Dates.now(), TimeZone(get_localzone()))
+end
```
Both of these could use the inline function syntax: `now() = ZonedDateTime...`

omus commented on src/timezones/local.jl:
```diff
@@ -0,0 +1,237 @@
+        @osx? _get_localzone_mac():(
+            _get_localzone_unix()
+        )
+    )
```
Switch to using `OS_NAME == :Windows` and `OS_NAME == :Darwin`

omus commented on src/timezones/local.jl:
```diff
@@ -0,0 +1,237 @@
+        end
+        tznames_raw = Array{UInt8}(tzh_charcnt)
+        namestart = 1
+        tznames = Dict{Uint8, String}()
```
Stick to using `UInt8` and don't mix with `Uint8`.

omus commented on src/timezones/local.jl:
```diff
@@ -0,0 +1,237 @@
+        for index in 1:tzh_charcnt
+            tznames_raw[index] = bswap(read(file, UInt8))
+            if tznames_raw[index] == '\0'
+                tznames[namestart] = ascii(tznames_raw[namestart:index-1])
```
If `index = 1` then you could have an exception being raised from `ascii([])`. Is this a problem?

iamed2 replied:
@omus Actually `ascii(UInt8[])` gives `""`

omus replied:
I stand corrected. Thanks

omus commented on src/timezones/local.jl:
```diff
@@ -0,0 +1,237 @@
+                namestart = index+1
+            end
+        end
+        timezone = nothing # The timezone to be returned
```
Don't mix types. Use `Nullable` instead.

omus replied:
Actually. Just return the timezone when you have it. But generally, don't mix types.

omus commented on src/timezones/local.jl:
```diff
@@ -0,0 +1,237 @@
+                else
+                    prev_inf = nothing
+                    for j in 0:i-1
+                        prev_inf = ttinfo[lindexes[i-j]]
```
I think this is much easier to understand:

```
for j in i:-1:1
    prev_inf = ttinfo[lindexes[j]]
```

omus commented on src/timezones/local.jl:
```diff
@@ -0,0 +1,237 @@
+                            continue
+                        end
+                    end
+                    dst = inf.tt_gmtoff - prev_inf.tt_gmtoff # dst offset
```
Not always true. What if the GMT offset for the zone changes for another reason?

omus commented on src/timezones/local.jl:
```diff
@@ -0,0 +1,237 @@
+    finally
+        close(file)
+    end
+    error("Failed to read timezone from tzinfo file")
```
In what case would this exception be raised?

iamed2 commented on src/timezones/local.jl:
```diff
@@ -0,0 +1,237 @@
+
+    filename = "/etc/timezone"
+    if isfile([filename]...)
+        file = open([filename]...)
```
Instead of the `try finally` pattern use `open( ) do` (in all cases where you've used `try finally` for files).

omus replied:
The `try finally` patter was being used since it appears to work better with mocking. But overall I agree.

omus committed:
Add appveyor support for Windows testing

Michael-Klassen committed:
Support 32-bit OS by using Number over Int64

omus commented on deps/build.jl:
```diff
@@ -43,3 +44,46 @@ end
 info("Successfully processed TimeZone data")
+
+
+if OS_NAME == :Windows #TODO change to window before commit
```
What's up with the comment? After talking with @iamed2  I think using `@windows_only begin` here is probably preferable.

omus commented on deps/build.jl:
```diff
@@ -43,3 +44,46 @@ end
+
+    # Add conversions to the dictionary
+    for mapzone in mapZones
+        if attribute(mapzone, "territory") == "001"
```
Why use territory "001"? Is there any documentation regarding this file you can link to?

omus replied:
We probably have to be more dynamic about this as some of the territory 001 entries don't exist but the alternative regions do. Here are the current unsupported timezone names:

```
(win,name) = ("India Standard Time","Asia/Calcutta")
(win,name) = ("UTC","Etc/GMT")
(win,name) = ("US Eastern Standard Time","America/Indianapolis")
(win,name) = ("Dateline Standard Time","Etc/GMT+12")
(win,name) = ("UTC-02","Etc/GMT+2")
(win,name) = ("Nepal Standard Time","Asia/Katmandu")
(win,name) = ("UTC-11","Etc/GMT+11")
(win,name) = ("Argentina Standard Time","America/Buenos_Aires")
(win,name) = ("UTC+12","Etc/GMT-12")
```

Michael-Klassen committed:
Create get_localzone support for Windows

Uses a downloaded conversion table to translate between Windows and
Olsen timezone names.

Michael-Klassen committed:
Added support for parsing a TimeZone from a tzfile

Michael-Klassen committed:
Added a limited number of fixed timezones

Michael-Klassen committed:
Renamed get_localzone to localzone

Really get_localzone was removed and all the OS specific implementations
all have identical signatures.

Michael-Klassen committed:
Improved accuracy in build_tzinfo

Michael-Klassen committed:
Create ZonedDateTime to DateTime conversion

omus committed:
Windows no longer creates an extra deps directory

omus committed:
Merge branch 'master' into localzone

Michael-Klassen committed:
Fix test using old localzone name

Michael-Klassen committed:
Fix a bug. GMT+i is a utc offset of -i hours and GMT-i is a utc offset of +i hours
